util/network/cniprovider/bridge.go

```diff
@@ -0,0 +1,150 @@
+package cniprovider
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	cni "github.com/containerd/go-cni"
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/moby/buildkit/util/network"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+)
+
+func NewBridge(opt Opt) (network.Provider, error) {
+	cniOptions := []cni.Opt{cni.WithInterfacePrefix("eth")}
+	bridgeBinName := "bridge"
+	loopbackBinName := "loopback"
+	hostLocalBinName := "host-local"
+	var setup bool
+	// binaries shipping with buildkit
+	for {
+		var dirs []string
+
+		bridgePath, err := exec.LookPath("buildkit-cni-bridge")
+		if err != nil {
+			break
+		}
+		var bridgeDir string
+		bridgeDir, bridgeBinName = filepath.Split(bridgePath)
+		dirs = append(dirs, bridgeDir)
+
+		loopbackPath, err := exec.LookPath("buildkit-cni-loopback")
+		if err != nil {
+			break
+		}
+		var loopbackDir string
+		loopbackDir, loopbackBinName = filepath.Split(loopbackPath)
+		if loopbackDir != bridgeDir {
+			dirs = append(dirs, loopbackDir)
+		}
+
+		hostLocalPath, err := exec.LookPath("buildkit-cni-host-local")
+		if err != nil {
+			break
+		}
+		var hostLocalDir string
+		hostLocalDir, hostLocalBinName = filepath.Split(hostLocalPath)
+		if hostLocalDir != bridgeDir && hostLocalDir != loopbackDir {
+			dirs = append(dirs, hostLocalDir)
+		}
+
+		cniOptions = append(cniOptions, cni.WithPluginDir(dirs))
+		setup = true
+		break //nolint: staticcheck
+	}
+
+	if !setup {
+		fn := filepath.Join(opt.BinaryDir, "bridge")
+		if _, err := os.Stat(fn); err != nil {
+			return nil, errors.Wrapf(err, "failed to find CNI bridge %q or buildkit-cni-bridge", fn)
+		}
+
+		cniOptions = append(cniOptions, cni.WithPluginDir([]string{opt.BinaryDir}))
+	}
+
+	cniOptions = append(cniOptions, cni.WithConfListBytes([]byte(fmt.Sprintf(`{
+		"cniVersion": "1.0.0",
+		"name": "buildkit",
+		"plugins": [
+			{
+				"type": "%s"
+			},
+			{
+				"type": "%s",
+				"bridge": "%s",
+				"isDefaultGateway": true,
+				"ipMasq": true,
+				"ipam": {
+				  "type": "%s",
+				  "ranges": [
+					[
+					  { "subnet": "10.10.0.0/16" }
+					]
+				  ]
+				}
+			  }
+		]
```

Comment:
Can we have `ingressPolicy": "same-bridge"` ?
https://www.cni.dev/plugins/current/meta/firewall/#ingress-policy

Reply:
So this would be another plugin we would need to include?

Reply:
Yes (`firewall` plugin)